### PR TITLE
[SVLS-8351] Add CPU Enhanced Metrics in Linux Azure Functions

### PR DESCRIPTION
**Pull request by kathiehuang**
### What does this PR do?
Adds CPU usage enhanced metric for Linux Azure Functions.
- Adds a new `datadog-metrics-collector` crate that reads CPU metrics every second and submits them to the Datadog backend every 10 seconds when `DD_ENHANCED_METRICS_ENABLED=true` (default on)
  - This creates an OS-agnostic `CpuMetricsCollector` struct and `CpuStatsReader` trait. Currently this only collects CPU metrics in Linux. CPU metrics in Windows will be completed in a future PR
- Emits one new metric:
  - `azure.functions.enhanced.cpu.usage` - container-level CPU consumption rate in nanocores, sourced from `cpuacct.usage`

### Additional Notes
- Tags attached to all CPU metrics: 
  - Azure resource metadata from `libdd-common`:
    - `resource_group`
    - `subscription_id`
    - `name`
  - Metadata from other environment variables:
    - `region`
    - `plan_tier`
    - `service`
    - `env`
    - `version`
    - `serverless_compat_version`
- Sets up CpuStats struct to store the computed CPU total
- Adds `DD_ENHANCED_METRICS` to `EnabledMetricsComponents` to determine which metrics components should be enabled
  - If `DD_USE_DOGSTATSD` is true (default true), start the DogStatsD listener
  - In Azure functions, the instance metrics collector always starts
  - If `DD_ENHANCED_METRICS` is true (default true in Azure Functions), start the CPU metrics collector
- Metrics are submitted as distribution metrics because not all metrics have tags with a unique identifier from the instance they are sent from
- If the collector cannot read the cgroup files successfully, it will not submit enhanced metrics for that interval and log accordingly

### Motivation
https://datadoghq.atlassian.net/browse/SVLS-8351

### Describe how to test/QA your changes
Build with serverless-compat-self-monitoring.

This was deployed with the serverless-compat-self-monitoring pipeline across all runtimes and hosting plans. All hosting plans in Linux were tested to verify metrics submit correctly, and a Windows function was tested to enhanced metrics are disabled in Windows environments.

Testing to compare against Azure Monitor is documented in an internal doc in [Enhanced Metrics in the Serverless Compatibility Layer](https://docs.google.com/document/d/1uMT4fphW7C31JLmf5KY1CdrXGPkn2QLfIvHZK3CJD40/edit?usp=sharing). Explanation of what cgroup contains can be found in [Calculating CPU Enhanced Metrics in Windows and Linux Azure Functions](https://datadoghq.atlassian.net/wiki/spaces/SLS/pages/6327238947/Calculating+CPU+Enhanced+Metrics+in+Windows+and+Linux+Azure+Functions).
* Verified CPU usage increases with a CPU-intensive function
* Verified CPU limit increases when the hosting plan is scaled up
* Verified enhanced metrics still send when `DD_USE_DOGSTATSD` is off
* Verified custom metrics still send when `DD_ENHANCED_METRICS_ENABLED` is off

References: [datadog-agent cgroup collection and calculation logic](https://github.com/DataDog/datadog-agent/blob/abab8cd7225c7a81fc72d1c7d43878a0d0816a8f/pkg/util/containers/metrics/system/collector_linux.go)

### CODE REVIEW

**Review by duncanista — COMMENTED** [review]
I'd suggest using features for OS specific business logic

Also suggest checking how ADP is doing agent checks in rust, this sounds like an agent check for a very specific use case

File: crates/datadog-serverless-compat/src/main.rs [hunk 1]
```diff
@@ -170,107 +180,103 @@ pub async fn main() {
...
-        )
-        .await;
-        if let Some(ref windows_pipe_name) = dd_dogstatsd_windows_pipe_name {
-            info!("dogstatsd-pipe: starting to listen on pipe {windows_pipe_name}");
+    let needs_aggregator = dd_use_dogstatsd || dd_enhanced_metrics;
+
+    let (metrics_flusher, aggregator_handle) = if needs_aggregator {
+        debug!("Creating metrics flusher and aggregator");
+
+        let (flusher, handle) =
```
Comment (duncanpharvey) [inline crates/datadog-serverless-compat/src/main.rs]: I think a comment here to note why the aggregator is started separately from the dogstatsd listener would be helpful - just enough to be clear that there are different configuration options that require this (dogstatsd enabled/disabled, enhanced metrics enabled/disabled).

Maybe a unit test as well to assert that all of these combinations are covered?

File: crates/datadog-serverless-compat/src/main.rs [hunk 2]
```diff
@@ -312,3 +361,42 @@ fn build_metrics_client(
     }
     Ok(builder.build()?)
 }
+
+fn build_cpu_metrics_tags() -> Option<SortedTags> {
```
Comment (duncanpharvey) [inline crates/datadog-serverless-compat/src/main.rs]: Would this method make more sense to live in the `datadog-metrics-collector` crate and be used internally within the crate?

**Review by duncanpharvey — COMMENTED** [review]
Excellent work! I added a few suggestions to consider

File: crates/datadog-serverless-compat/src/main.rs (hunk 1 above)
Comment (kathiehuang) [reply crates/datadog-serverless-compat/src/main.rs]: Good point! I added a comment in https://github.com/DataDog/serverless-components/pull/77/commits/60cdecf682ea06d1c45b3226f1eaedfd6bb39ad1

It seems like it'll be a little hard to make a meaningful unit test since the aggregator/dogstatsd startup logic has side effects that would make it hard to test in isolation? Maybe I could refactor the startup decision into a struct that describes what to start to separate the decision from execution?


```
struct AgentConfig {
    start_aggregator: bool,
    start_dogstatsd: bool,
    start_enhanced_metrics: bool,
}

fn resolve_agent_config(dd_use_dogstatsd: bool, dd_enhanced_metrics: bool) -> AgentConfig {
    AgentConfig {
        start_aggregator: dd_use_dogstatsd || dd_enhanced_metrics,
        start_dogstatsd: dd_use_dogstatsd,
        start_enhanced_metrics: dd_enhanced_metrics,
    }
}
```

File: crates/datadog-serverless-compat/src/main.rs (hunk 2 above)
Comment (kathiehuang) [reply crates/datadog-serverless-compat/src/main.rs]: That makes sense! I moved it in https://github.com/DataDog/serverless-components/pull/77/commits/f867f6f1ee8ca942f21c840934a11839614d0603

File: crates/datadog-metrics-collector/src/azure_cpu.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
+// Copyright 2023-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+//! CPU metrics collector for Azure Functions
```
[the remark this answers is not in the document]
Comment (kathiehuang) [reply crates/datadog-metrics-collector/src/azure_cpu.rs:4]: Since this crate is only used in Azure Functions environments right now, I'm thinking of leaving the name as-is and renaming it in the future if we expand enhanced metrics to Cloud Functions Gen 1
Edit: NVM, renamed in https://github.com/DataDog/serverless-components/pull/77/commits/ae8b3c7c088049977112a7ae5f85692d2ca1b7a1 to make it clear that these are Azure-specific without having to open up the file

File: crates/datadog-metrics-collector/src/cpu.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+impl CpuMetricsCollector {
+    /// Creates a new CpuMetricsCollector
+    ///
+    /// # Arguments
+    ///
+    /// * `aggregator` - The aggregator handle to submit metrics to
+    /// * `tags` - Optional tags to attach to all metrics
+    pub fn new(aggregator: AggregatorHandle, tags: Option<SortedTags>) -> Self {
+        #[cfg(feature = "windows-enhanced-metrics")]
```
Comment (Lewis-E) [inline crates/datadog-metrics-collector/src/cpu.rs]: Nit-flavored, but: what about `#[cfg(all(windows, feature = "windows-enhanced-metrics"))]` and then `#[cfg(not(windows))]` on line 49?

File: crates/datadog-serverless-compat/src/main.rs [hunk 5]
```diff
@@ -170,107 +184,106 @@ pub async fn main() {
...
+                if let Some(metrics_flusher) = metrics_flusher.clone() {
+                    debug!("Flushing dogstatsd metrics");
+                    tokio::spawn(async move {
+                        metrics_flusher.flush().await;
+                    });
+                }
+            }
+            _ = cpu_collection_interval.tick() => {
+                if let Some(ref mut collector) = cpu_collector {
+                    collector.collect_and_submit();
```
Comment (Lewis-E) [inline crates/datadog-serverless-compat/src/main.rs:370]: The go agent now has `WaitForPendingSamples` as a drain on shutdown. It might be good to have a followup ticket to add that behavior as one more select arm? I don't think this needs to block, and in fact I think it will be better reviewed separately, but shutdown handling may contribute to incorrect/mismatched metrics.

File: crates/datadog-serverless-compat/src/main.rs [hunk 6]
```diff
@@ -104,6 +107,17 @@ pub async fn main() {
...
 
+    // Only enable enhanced metrics for Linux Azure Functions
+    #[cfg(not(feature = "windows-enhanced-metrics"))]
+    let dd_enhanced_metrics = env_type == EnvironmentType::AzureFunction
+        && env::var("DD_ENHANCED_METRICS_ENABLED")
+            .map(|val| val.to_lowercase() != "false")
+            .unwrap_or(true);
+
+    // Enhanced metrics are not yet supported in Windows environments
+    #[cfg(feature = "windows-enhanced-metrics")]
```
[the remark this answers is not in the document]
Comment (Lewis-E) [reply crates/datadog-serverless-compat/src/main.rs]: (I think explicit gating is a good idea and tried to do it in the other windows feature)

**Review by Lewis-E — APPROVED** [review]

File: crates/datadog-metrics-collector/src/cpu.rs (hunk 4 above)
Comment (kathiehuang) [reply crates/datadog-metrics-collector/src/cpu.rs]: That makes sense and would make sure this feature doesn't accidentally get built on non-Windows environments! Updated in https://github.com/DataDog/serverless-components/pull/77/commits/3775868825d7d18188d14cec976e8d4320cf93aa

File: crates/datadog-serverless-compat/src/main.rs (hunk 5 above)
Comment (kathiehuang) [reply crates/datadog-serverless-compat/src/main.rs:370]: Good point! Right now in Azure Functions we don't have a shutdown signal but this would definitely be a good follow up. I think tailing the logs could be a way to detect when a function is shutting down so that we can drain the queue. Made a ticket to do this as follow-up https://datadoghq.atlassian.net/browse/SVLS-9026

File: crates/datadog-serverless-compat/src/main.rs [hunk 7]
```diff
@@ -117,6 +120,17 @@ pub async fn main() {
...
+    #[cfg(not(windows))]
+    let dd_enhanced_metrics = env_type == EnvironmentType::AzureFunction
+        && env::var("DD_ENHANCED_METRICS_ENABLED")
+            .map(|val| val.to_lowercase() != "false")
+            .unwrap_or(true);
+
+    // Enhanced metrics are not yet supported in Windows environments
+    #[cfg(all(windows, feature = "windows-enhanced-metrics"))]
+    let dd_enhanced_metrics = false;
+
```
[the remark this answers is not in the document]
Comment (kathiehuang) [reply crates/datadog-serverless-compat/src/main.rs:133]: This will not happen - we handle building and publishing

File: crates/datadog-serverless-compat/src/main.rs [hunk 8]
```diff
@@ -292,8 +324,11 @@ pub async fn main() {
     let mut instance_metrics_collection_interval = interval(Duration::from_secs(
         INSTANCE_METRICS_COLLECTION_INTERVAL_SECS,
     ));
+    let mut cpu_collection_interval =
+        interval(Duration::from_secs(CPU_METRICS_COLLECTION_INTERVAL_SECS));
```
[the remark this answers is not in the document]
Comment (kathiehuang) [reply crates/datadog-serverless-compat/src/main.rs:328]: Good catch, updated in https://github.com/DataDog/serverless-components/pull/77/commits/b84dc84ffbd9a4df8bf678a4fd869b1d3eb929e5 to check if `cpu_collector` is None to conditionally wake up the loop

File: crates/datadog-metrics-collector/src/azure_cpu.rs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                .duration_since(self.last_collection_time)
+                .as_secs_f64();
+            self.last_collection_time = now_instant;
+            if elapsed_secs <= 0.0 {
+                debug!("Elapsed time is less than or equal to 0, skipping interval");
+                return;
+            }
+
+            // Divide nanoseconds delta by elapsed time to get usage rate in nanocores
+            let usage_rate_nc = delta_ns / elapsed_secs;
```
[the remark this answers is not in the document]
Comment (kathiehuang) [reply crates/datadog-metrics-collector/src/azure_cpu.rs]: The math is pretty straightforward and the edge cases are explicit, but I'm happy to add unit tests if anyone agrees here!

File: crates/datadog-serverless-compat/src/main.rs [hunk 10]
```diff
@@ -117,6 +120,17 @@ pub async fn main() {
...
 
+    // Only enable enhanced metrics for Linux Azure Functions
+    #[cfg(not(windows))]
+    let dd_enhanced_metrics = env_type == EnvironmentType::AzureFunction
+        && env::var("DD_ENHANCED_METRICS_ENABLED")
+            .map(|val| val.to_lowercase() != "false")
+            .unwrap_or(true);
+
+    // Enhanced metrics are not yet supported in Windows environments
+    #[cfg(all(windows, feature = "windows-enhanced-metrics"))]
```
Comment (duncanpharvey) [inline crates/datadog-serverless-compat/src/main.rs:131]: The `windows-enhanced-metrics` feature can be omitted here. By adding it there is a gap if this is compiled for windows without the `windows-enhanced-metrics` feature. That's not done in practice right now but it does make the code more fragile having this gap.

File: crates/datadog-metrics-collector/src/azure_cpu.rs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                .map(|d| d.as_secs())
+                .unwrap_or(0)
+                .try_into()
+                .unwrap_or(0);
+
+            let usage_metric = Metric::new(
+                CPU_USAGE_METRIC.into(),
+                MetricValue::distribution(usage_rate_nc),
+                self.tags.clone(),
+                Some(now),
```
Comment (duncanpharvey) [inline crates/datadog-metrics-collector/src/azure_cpu.rs]: Could we use `now_instant` as the timestamp on the metric since that's when the usage data point was collected?

File: crates/datadog-metrics-collector/src/azure_cpu.rs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            reader,
+            aggregator,
+            tags,
+            last_usage_ns: None,
+            last_collection_time: std::time::Instant::now(),
+        }
+    }
+
+    pub fn collect_and_submit(&mut self) {
+        if let Some(cpu_stats) = self.reader.read() {
```
Comment (duncanpharvey) [inline crates/datadog-metrics-collector/src/azure_cpu.rs]: nit: could flatten this large if/else block with a small refactor

```
      let Some(cpu_stats) = self.reader.read() else {
          debug!("Skipping CPU enhanced metrics collection - could not find data to generate CPU usage metrics");
          return;
      };
```

**Review by duncanpharvey — APPROVED** [review]
Looks good! just a few nits and minor comments. Could you also add some unit tests to cover the functionality in the cpu stats reader and the metric collector?

File: crates/datadog-serverless-compat/src/main.rs (hunk 10 above)
Comment (kathiehuang) [reply crates/datadog-serverless-compat/src/main.rs:131]: Here are earlier conversations about the Windows feature for context!

> I'd suggest using features for OS specific business logic

@duncanista

> Nit-flavored, but: what about `#[cfg(all(windows, feature = "windows-enhanced-metrics"))]` and then `#[cfg(not(windows))]` on line 49?

@Lewis-E

and there is a precedent for using features for Windows-specific logic with `windows-pipes`

File: crates/datadog-metrics-collector/src/azure_cpu.rs (hunk 11 above)
Comment (kathiehuang) [reply crates/datadog-metrics-collector/src/azure_cpu.rs]: I can't use `now_instant` directly in this function because `Instant` is a monotonic clock with no fixed relationship to Unix time, but I moved the creation of `now` to be right after the creation of `now_instant` in https://github.com/DataDog/serverless-components/pull/77/commits/24924b09ebd6d62d255179886e33bb2b7450194a!

File: crates/datadog-metrics-collector/src/azure_cpu.rs (hunk 12 above)
Comment (kathiehuang) [reply crates/datadog-metrics-collector/src/azure_cpu.rs]: True, that would reduce cyclomatic complexity. Updated in https://github.com/DataDog/serverless-components/pull/77/commits/0430a664a4b97ab854024020169fb42a85bf0e00!

**Review by duncanpharvey — APPROVED** [review]
